### PR TITLE
Add batch add interface to serial subsystem

### PR DESCRIPTION
**Pull request by wom-bat**
Add many characters to a queue using memcpy()

### CODE REVIEW

**Comment by Ivan-Velickovic** [comment]
Should we be taking in a `uint8_t *` instead? This function seems useful if you are not dealing with strings/text data as well.

**Comment by Ivan-Velickovic** [comment]
And instead of talking about 'characters' we just refer to 'bytes' since the serial system should not care kind of data it is.

**Comment by Courtney3141** [comment]
@Ivan-Velickovic  changed the description to bytes rather than characters. Not sure about the type change so I'll leave that one to @wom-bat

File: include/sddf/serial/queue.h [hunk 1]
```diff
@@ -218,6 +218,29 @@ static inline uint32_t serial_queue_contiguous_free(serial_queue_handle_t *queue
...
+    uint32_t avail = serial_queue_contiguous_free(qh);
+    char *p = qh->data_region + (qh->queue->tail % qh->size);
+    n =  MIN(n, avail);
+    sddf_memcpy(p, src, n);
+
+    serial_update_visible_tail(qh, qh->queue->tail + n);
+
+    return n;
+}
+
```
[the remark this answers is not in the document]
Comment (wom-bat) [reply include/sddf/serial/queue.h:243]: I like this.  I thought about doing it initially, but wanted to keep things simple to start with.

The called still needs to deal with a partial write.

[the remark this answers is not in the document]
Comment (wom-bat) [reply include/sddf/serial/queue.h:243]: Plus calling `MIN` on anything other than simple variables or constants is fraught without a more complex implementation of `MIN` --- it'll invoke `serial_queue_contiguous_free()` twice.

**Comment by wom-bat** [comment]
> @Ivan-Velickovic changed the description to bytes rather than characters. Not sure about the type change so I'll leave that one to @wom-bat

Serial works in characters (which could be 6, 7 or 8 bits each), not bytes.

**Comment by Ivan-Velickovic** [comment]
> Serial works in characters (which could be 6, 7 or 8 bits each), not bytes.

It seems weird then for our interface to suddenly switch depending on whatever people have defined as `char`. Shouldn't we always be transferring a set number of bits?

**Comment by wom-bat** [comment]
> > Serial works in characters (which could be 6, 7 or 8 bits each), not bytes.
> 
> It seems weird then for our interface to suddenly switch depending on whatever people have defined as `char`. Shouldn't we always be transferring a set number of bits?

In 'C' char is always 8 bits (might be signed or unsigned depending on platform).    But in terms of what the hardware spits out or reads, only some of them will be used.
Keeping the interface as 'char' is meant to remind people of this.

**Comment by Ivan-Velickovic** [comment]
Sure, maybe we should drop https://github.com/au-ts/sddf/pull/159/commits/9692fcd9161d4e3d390f9c315a1f894e2b0a480e then.

**Review by Courtney3141 — DISMISSED** [review]

**Review by Ivan-Velickovic — APPROVED** [review]

**Pull request merged**